cloudinit/sources/DataSourceVultr.py

```diff
@@ -0,0 +1,136 @@
+# Author: Eric Benner <ebenner@vultr.com>
+#
+# This file is part of cloud-init. See LICENSE file for license information.
+
+# Vultr Metadata API:
+# https://www.vultr.com/metadata/
+
+import json
+
+from cloudinit import log as log
+from cloudinit import sources
+from cloudinit import util
+
+import cloudinit.sources.helpers.vultr as vultr
+
+LOG = log.getLogger(__name__)
+BUILTIN_DS_CONFIG = {
+    'url': 'http://169.254.169.254',
+    'retries': 30,
+    'timeout': 2,
+    'wait': 2
+}
+
+
+class DataSourceVultr(sources.DataSource):
+
+    dsname = 'Vultr'
+
+    def __init__(self, sys_cfg, distro, paths):
+        super(DataSourceVultr, self).__init__(sys_cfg, distro, paths)
+        self.ds_cfg = util.mergemanydict([
+            util.get_cfg_by_path(sys_cfg, ["datasource", "Vultr"], {}),
+            BUILTIN_DS_CONFIG])
+        BUILTIN_DS_CONFIG['url'] = self.ds_cfg.get(
+            'url', BUILTIN_DS_CONFIG['url'])
+        BUILTIN_DS_CONFIG['retries'] = self.ds_cfg.get(
+            'retries', BUILTIN_DS_CONFIG['retries'])
+        BUILTIN_DS_CONFIG['timeout'] = self.ds_cfg.get(
+            'timeout', BUILTIN_DS_CONFIG['timeout'])
+        BUILTIN_DS_CONFIG['wait'] = self.ds_cfg.get(
+            'wait', BUILTIN_DS_CONFIG['wait'])
+
+    # Initiate data and check if Vultr
+    def _get_data(self):
+        LOG.debug("Detecting if machine is a Vultr instance")
+        if not vultr.is_vultr():
+            LOG.debug("Machine is not a Vultr instance")
+            return False
+
+        LOG.debug("Machine is a Vultr instance")
+
+        config = vultr.generate_config(self.ds_cfg)
+
+        # Dump vendor config so diagnosing failures is manageable
+        LOG.debug("Vultr Vendor Config:")
+        LOG.debug(json.dumps(config))
+
+        md = self.get_metadata()
+
+        self.metadata_full = md
+        self.metadata['instanceid'] = self.metadata_full['instanceid']
+        self.metadata['local-hostname'] = self.metadata_full['hostname']
+
+        # Default hostname is "vultr"
+        if self.metadata['local-hostname'] == "":
+            self.metadata['local-hostname'] = "vultr"
+
+        self.metadata['public-keys'] = md["public-keys"].splitlines()
+        self.userdata_raw = md["user-data"]
+        if self.userdata_raw == "":
+            self.userdata_raw = None
+        self.vendordata_raw = "#cloud-config\n%s" % json.dumps(config)
```

Comment:
yeah... so as i read/thought about this more, this should be bytes i think.  sorry that you exposed some inconsistency in cloud-init.

Reply:
So to confirm you want me to change this to bytes and close the following PR?

https://github.com/canonical/cloud-init/pull/837